### PR TITLE
Receive and send RST frames

### PR DESCRIPTION
**Pull request by Tratcher**
#2462 @halter73 I still need to finish the logging and error messages, but I wanted to show you the concept first.
Once the client has sent us a reset for a given stream we should stop sending headers or data on that stream. The request continues normally and OnStarting still triggers, you only see the error if you check RequestAborted or pass a cancellable token to WriteAsync.

### CODE REVIEW

**Comment by Tratcher** [comment]
@halter73 getting the app to send RST when it calls Abort wasn't as bad as I thought. This should be additive on top of your changes.

**Comment by Tratcher** [comment]
Updated: This finishes off receiving and sending RST frames.

File: test/Kestrel.Core.Tests/Http2ConnectionTests.cs [hunk 1]
```diff
@@ -2312,6 +2306,104 @@ public Task HEADERS_Received_HeaderBlockContainsTEHeader_ValueIsNotTrailers_Stre
...
+                    _runningStreams[streamIdFeature.StreamId].TrySetException(ex);
+                }
+            });
+
+            await StartStreamAsync(1, _browserRequestHeaders, endStream: true);
+            await WaitForStreamErrorAsync(expectedStreamId: 1, Http2ErrorCode.CANCEL, expectedErrorMessage: null);
+            await WaitForAllStreamsAsync();
+            Assert.Contains(1, _abortedStreamIds);
+
+            await SendGoAwayAsync();
```
Comment (halter73) [inline test/Kestrel.Core.Tests/Http2ConnectionTests.cs]: Why not just StopConnectionAsync? That should work, right?

File: test/Kestrel.Core.Tests/Http2ConnectionTests.cs [hunk 2]
```diff
@@ -2312,6 +2306,104 @@ public Task HEADERS_Received_HeaderBlockContainsTEHeader_ValueIsNotTrailers_Stre
...
+            await ExpectAsync(Http2FrameType.DATA,
+                withLength: 10,
+                withFlags: 0,
+                withStreamId: 1);
+
+            await WaitForStreamErrorAsync(expectedStreamId: 1, Http2ErrorCode.CANCEL, expectedErrorMessage: null);
+            await WaitForAllStreamsAsync();
+            Assert.Contains(1, _abortedStreamIds);
+
+            await SendGoAwayAsync();
```
Comment (halter73) [inline test/Kestrel.Core.Tests/Http2ConnectionTests.cs]: StopConnectionAsync

File: test/Kestrel.Core.Tests/Http2ConnectionTests.cs [hunk 3]
```diff
@@ -1932,8 +1928,6 @@ public Task HEADERS_Received_HeaderBlockContainsTEHeader_ValueIsNotTrailers_Stre
 
             // No END_STREAM HEADERS or DATA frame is received since the stream was aborted
             await WaitForConnectionStopAsync(expectedLastStreamId: 1, ignoreNonGoAwayFrames: false);
-
-            // TODO: Check logs
```
Comment (halter73) [inline test/Kestrel.Core.Tests/Http2ConnectionTests.cs]: I noticed these tests call `SendGoAwayAsync` too. Prior to these being added, only the `GOAWAY_` tests called this helper. Any other test that wanted to close the stream just closed the pipe. I think we should keep this consistent.

File: src/Kestrel.Core/Internal/Http2/Http2Stream.cs [hunk 4]
```diff
@@ -144,17 +146,51 @@ protected override bool TryParseRequest(ReadResult result, out bool endConnectio
...
-        public bool TryUpdateOutputWindow(int bytes)
+        protected override void ApplicationAbort()
         {
-            return _context.FrameWriter.TryUpdateStreamWindow(_outputFlowControl, bytes);
+            Log.ApplicationAbortedConnection(ConnectionId, TraceIdentifier);
+            var abortReason = new ConnectionAbortedException(CoreStrings.ConnectionAbortedByApplication);
+            AbortAndReset(abortReason, Http2ErrorCode.CANCEL);
+        }
+
+        public void AbortAndReset(ConnectionAbortedException abortReason, Http2ErrorCode error)
```
Comment (halter73) [inline src/Kestrel.Core/Internal/Http2/Http2Stream.cs]: private

File: src/Kestrel.Core/Internal/Http2/Http2OutputProducer.cs [hunk 5]
```diff
@@ -46,6 +46,8 @@ public class Http2OutputProducer : IHttpOutputProducer
             _dataWriteProcessingTask = ProcessDataWrites();
         }
 
+        public bool IsCompleted => _completed;
```
Comment (halter73) [inline src/Kestrel.Core/Internal/Http2/Http2OutputProducer.cs]: It's not safe to check `_completed` outside the lock. I see that the way it's being used now is basically an optimization, but `HttpContext.Abort()` is not called frequently enough for this to be worth it. Remove this property.

File: src/Kestrel.Core/Internal/Http2/Http2Stream.cs [hunk 6]
```diff
@@ -144,17 +146,51 @@ protected override bool TryParseRequest(ReadResult result, out bool endConnectio
...
+        {
+            if (Interlocked.Exchange(ref _requestAborted, 1) != 0)
+            {
+                return;
+            }
+
+            var outputCompleted = Output.IsCompleted;
```
Comment (halter73) [inline src/Kestrel.Core/Internal/Http2/Http2Stream.cs]: Just put WriteRstStreamAsync on the OutputProducer to avoid any potential race conditions.

File: test/Kestrel.Core.Tests/Http2ConnectionTests.cs (hunk 3 above)
Comment (Tratcher) [reply test/Kestrel.Core.Tests/Http2ConnectionTests.cs]: GO_AWAY is relevant to the RST tests since it calls the Abort code path.

File: test/Kestrel.Core.Tests/Http2ConnectionTests.cs [hunk 7]
```diff
@@ -529,10 +529,10 @@ void IHttpHeadersHandler.OnHeader(Span<byte> name, Span<byte> value)
             await StartStreamAsync(1, _browserRequestHeaders, endStream: false);
             await StartStreamAsync(3, _browserRequestHeaders, endStream: false);
 
-            await SendDataAsync(1, _helloBytes, endStream: false);
+            await SendDataAsync(1, _helloBytes, endStream: true);
             Assert.True(stream1Read.WaitOne(TimeSpan.FromSeconds(10)));
 
-            await SendDataAsync(3, _helloBytes, endStream: false);
+            await SendDataAsync(3, _helloBytes, endStream: true);
```
Comment (halter73) [inline test/Kestrel.Core.Tests/Http2ConnectionTests.cs]: Was this change necessary for the test to pass?

Comment (Tratcher) [reply test/Kestrel.Core.Tests/Http2ConnectionTests.cs]: Leftover from a prior version, reverted.

File: test/Kestrel.Core.Tests/Http2ConnectionTests.cs (hunk 3 above)
Comment (halter73) [reply test/Kestrel.Core.Tests/Http2ConnectionTests.cs]: I don't buy it. If it was relevant, these tests wouldn't still pass when removing the SendGoAwayAsync() call. 

The ProcessGoAwayFrameAsync() just cancels the pending read on the Http2Connection's Input PipeReader stopping the ProcessRequestsAsync() loop. Completing the PipeWriter on the other side via StopConnectionAsync() also stops the loop and aborts all the streams.

**Comment by Tratcher** [comment]
Updated

File: src/Kestrel.Core/Internal/Http2/Http2OutputProducer.cs [hunk 8]
```diff
@@ -181,6 +180,22 @@ public Task WriteStreamSuffixAsync()
             }
         }
 
+        public Task WriteRstStreamAsync(bool endReceived, Http2ErrorCode error)
+        {
+            lock (_dataWriterLock)
+            {
+                // Only reset if the connection is open or in a half closed state.
+                if (endReceived && _completed)
```
Comment (halter73) [inline src/Kestrel.Core/Internal/Http2/Http2OutputProducer.cs]: I don't like evaluating endReceived in the lock. It really implies that Http2Stream.EndStreamReceived is protected by this lock, but it's not.

File: src/Kestrel.Core/Internal/Http2/Http2OutputProducer.cs [hunk 9]
```diff
@@ -181,6 +180,22 @@ public Task WriteStreamSuffixAsync()
...
+                {
+                    return Task.CompletedTask;
+                }
+
+                _completed = true;
```
Comment (halter73) [inline src/Kestrel.Core/Internal/Http2/Http2OutputProducer.cs]: If you set `_completed` to true, you're responsible for completing the data PipeWriter. I would probably just call Dispose() here.

File: src/Kestrel.Core/Internal/Http2/Http2Stream.cs [hunk 10]
```diff
@@ -144,17 +147,45 @@ protected override bool TryParseRequest(ReadResult result, out bool endConnectio
...
 
             // Unblock the request body.
             RequestBodyPipe.Writer.Complete(new IOException(CoreStrings.Http2StreamAborted, abortReason));
         }
 
-        public bool TryUpdateOutputWindow(int bytes)
+        protected override void ApplicationAbort()
         {
-            return _context.FrameWriter.TryUpdateStreamWindow(_outputFlowControl, bytes);
+            Log.ApplicationAbortedConnection(ConnectionId, TraceIdentifier);
```
Comment (halter73) [inline src/Kestrel.Core/Internal/Http2/Http2Stream.cs]: Should this be logged if the stream was already aborted? Why isn't this combined with ResetAndAbort()?

File: src/Kestrel.Core/Internal/Http2/Http2Stream.cs [hunk 11]
```diff
@@ -144,17 +147,45 @@ protected override bool TryParseRequest(ReadResult result, out bool endConnectio
...
+
+        private void ResetAndAbort(ConnectionAbortedException abortReason, Http2ErrorCode error)
+        {
+            if (Interlocked.Exchange(ref _requestAborted, 1) != 0)
+            {
+                return;
+            }
+
+            // If this was an app initiated Abort then we need to inform the client.
+            _http2Output.WriteRstStreamAsync(endReceived: EndStreamReceived, error).GetAwaiter().GetResult();
```
Comment (halter73) [inline src/Kestrel.Core/Internal/Http2/Http2Stream.cs]: Having `HttpContext.Abort()` block on I/O for any reason is new. I don't think we should do it. Fire and forget is better. This write can't fail since no cancellation token is passed in anyway.

File: src/Kestrel.Core/Internal/Http2/Http2Stream.cs [hunk 12]
```diff
@@ -144,17 +147,45 @@ protected override bool TryParseRequest(ReadResult result, out bool endConnectio
...
+
+            base.Abort(abortReason);
+
+            // Unblock the request body.
+            RequestBodyPipe.Writer.Complete(new IOException(CoreStrings.Http2StreamAborted, abortReason));
```
Comment (halter73) [inline src/Kestrel.Core/Internal/Http2/Http2Stream.cs]: Maybe put the above 2 lines in a shared AbortCore() method.

File: src/Kestrel.Core/Internal/Http2/Http2Stream.cs (hunk 10 above)
Comment (Tratcher) [reply src/Kestrel.Core/Internal/Http2/Http2Stream.cs]: I separated ResetAndAbort because I expect it to be called for other conditions like my next PR for invalid request data. There will be different log statements.

Comment (halter73) [reply src/Kestrel.Core/Internal/Http2/Http2Stream.cs]: I'd combine the methods until there's another caller, but no big deal.

I'm not sure I get the point about there being different log statements. I'm just pointing out it seems wrong to log that app aborted the stream if the stream was already aborted.

Comment (Tratcher) [reply src/Kestrel.Core/Internal/Http2/Http2Stream.cs]: True, ApplicationAbortedConnection could be logged multiple times, but only for calls to HttpContext.Abort, not for any other reason. That's actually the same as it used to be if you look at the base class (HTTP/1.1) behavior we're overriding. Redundant but harmless.

File: src/Kestrel.Core/Internal/Http2/Http2OutputProducer.cs (hunk 8 above)
Comment (halter73) [reply src/Kestrel.Core/Internal/Http2/Http2OutputProducer.cs]: @Tratcher pointed out this is an `&&` not `||` offline, so now I see why it was done this way. We decided to send the RST regardless since that's allowed by spec. The server/app shouldn't be likely to try to abort a dead stream anyway.

Comment (halter73) [reply src/Kestrel.Core/Internal/Http2/Http2OutputProducer.cs]: Just add a comment saying that it was intentional that we don't take the lock or check completed in this method.

File: src/Kestrel.Core/Internal/Http2/Http2Stream.cs (hunk 10 above)
Comment (halter73) [reply src/Kestrel.Core/Internal/Http2/Http2Stream.cs]: OK

**Comment by Tratcher** [comment]
Updated.

File: src/Kestrel.Core/Internal/Http2/Http2Stream.cs [hunk 13]
```diff
@@ -144,17 +147,46 @@ protected override bool TryParseRequest(ReadResult result, out bool endConnectio
...
+        }
+
+        private void ResetAndAbort(ConnectionAbortedException abortReason, Http2ErrorCode error)
+        {
+            if (Interlocked.Exchange(ref _requestAborted, 1) != 0)
+            {
+                return;
+            }
+
+            var ignored = _http2Output.WriteRstStreamAsync(error);
```
Comment (halter73) [inline src/Kestrel.Core/Internal/Http2/Http2Stream.cs]: Nit: `_ =`

Maybe comment saying the task can't fault.
